src/simulation.F90

```diff
@@ -498,9 +504,18 @@ subroutine openmc_simulation_finalize() bind(C)
     ! Broadcast tally results so that each process has access to results
     if (allocated(tallies)) then
       do i = 1, size(tallies)
-        n = size(tallies(i) % obj % results)
-        call MPI_BCAST(tallies(i) % obj % results, n, MPI_DOUBLE, 0, &
-             mpi_intracomm, mpi_err)
+        associate (results => tallies(i) % obj % results)
+          ! Create a new datatype that consists of all values for a given filter
+          ! bin and then use that to broadcast. This is done to minimize the
+          ! chance of the 'count' argument of MPI_BCAST exceeding 2**31
+          n = size(results, 3)
+          count_per_filter = size(results, 1) * size(results, 2)
+          call MPI_TYPE_CONTIGUOUS(count_per_filter, MPI_DOUBLE, &
+               result_block, mpi_err)
```

Comment:
The tally results array is actually declared to be of type [double](https://github.com/mit-crpg/openmc/blob/develop/src/tallies/tally_header.F90#L90) for the purpose of C interoperability, so that's why `MPI_DOUBLE` was used. In practical terms, a `real(8)` ends up being the same as a `real(C_DOUBLE)`.